paradox/hardware/spectra_magellan/panel.py

```diff
@@ -0,0 +1,257 @@
+# -*- coding: utf-8 -*-
+
+import inspect
+import sys
+import logging
+import time
+from .parsers import *
+from ..panel import Panel as PanelBase
+from config import user as cfg
+
+MEM_STATUS_BASE1 = 0x8000
+MEM_STATUS_BASE2 = 0x1fe0
+
+MEM_ZONE_START = 0x010
+MEM_ZONE_END = MEM_ZONE_START + 0x10 * 32
+MEM_OUTPUT_START = MEM_ZONE_END
+MEM_OUTPUT_END = MEM_OUTPUT_START + 0x10 * 16
+MEM_PARTITION_START = MEM_OUTPUT_END
+MEM_PARTITION_END = MEM_PARTITION_START + 0x10 * 2
+MEM_USER_START = MEM_PARTITION_END
+MEM_USER_END = MEM_USER_START + 0x10 * 32
+MEM_BUS_START = MEM_USER_END
+MEM_BUS_END = MEM_BUS_START + 0x10 * 15
+MEM_REPEATER_START = MEM_BUS_END
+MEM_REPEATER_END = MEM_REPEATER_START + 0x10 * 2
+MEM_KEYPAD_START = MEM_REPEATER_END
+MEM_KEYPAD_END = MEM_KEYPAD_START + 0x10 * 8
+MEM_SITE_START = MEM_KEYPAD_END
+MEM_SITE_END = MEM_SITE_START + 0x10
+MEM_SIREN_START = MEM_SITE_END
+MEM_SIREN_END = MEM_SIREN_START + 0x10 * 4
+
+logger = logging.getLogger('PAI').getChild(__name__)
+
+
+class Panel(PanelBase):
+    def get_message(self, name):
+        try:
+            return super(Panel, self).get_message(name)
+        except ResourceWarning as e:
+            clsmembers = dict(inspect.getmembers(sys.modules[__name__]))
+            if name in clsmembers:
+                return clsmembers[name]
+            else:
+                raise e
+
+    def update_labels(self):
```

Comment:
If we have common things in the different panels, a better approach would be to have a structure with the addresses, and put the common elements in the parent class. At least zones and partition labels will be common.

Reply:
EVO192 zones are divided in two non continuous address ranges. Will need to add support for an array of ranges.

Partition names are in some structure where each next partition label is after 107 bytes from beginning of previous. But label length is always 16 bytes. Probably _size_ is bad name in _mem_map_ then. It is not _size_, but _shift_ or _offset_ ?
Same thing for PGMs.